### PR TITLE
feat: support modern mode

### PR DESCRIPTION
**Pull request by pigcan**
##### Checklist

- [x] `npm test` passes
- [x] tests are included
- [x] documentation is changed or added
- [x] commit message follows commit guidelines

##### Description of change

- any feature?

Introduce modern mode 

[RFC - Modern Mode](https://github.com/umijs/rfcs/pull/5)

### CODE REVIEW

**Comment by pigcan** [comment]
用例现在还没挂起钩，之后我再处理下吧

**Comment by sorrycc** [comment]
beforeProdCompileAsync 改成 beforeBuildCompileAsync，和 `onBuildSuccess` 等保持命名一致。

File: packages/umi-build-dev/src/plugins/commands/build/index.js [hunk 1]
```diff
@@ -36,43 +36,46 @@ export default function(api) {
...
-              }
-              service.applyPlugins('onBuildSuccess', {
-                args: {
-                  stats,
-                },
-              });
-              service
-                ._applyPluginsAsync('onBuildSuccessAsync', {
+          service._applyPluginsAsync('beforeBuildCompileAsync').then(() => {
+            console.log('haha');
```
Comment (afc163) [inline packages/umi-build-dev/src/plugins/commands/build/index.js]: haha

Comment (pigcan) [reply packages/umi-build-dev/src/plugins/commands/build/index.js]: 😄 @sorrycc

**Pull request merged**